Implement multiuser login

chanseokoh opened the pull request:
However, functionality-wise, everything is working, including adding and switching new accounts. The existing deploy flow works. If you're curious, you can check out this branch and run Eclipse to test them.

UI is not pretty (which is my best):

![java - nothing-src-main-java-something-nothing-helloappengine java - eclipse platform _014](https://cloud.githubusercontent.com/assets/10523105/18852445/6c51e032-840e-11e6-8bb9-0a8c42d57cc2.png)

The accounts panel is shown only when there is at least one account signed in. The menu item and the icon tooltip will also change based on the login status.

Things to do:
 - Add as many tests as possible, in this PR.
 - Move the accounts panel to the bottom-right corner, in another PR. (#740)

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/plugin.properties:
```diff
@@ -1,2 +1,3 @@
-providerName=Google, Inc.
\ No newline at end of file
+providerName=Google, Inc.
+loginIconTooltip=Sign in with your Google Account to integrate with Google services
```
integrate with --> access

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -96,42 +111,108 @@ public Shell getShell() {
+      Account account = findAccount(accounts, email);
+      if (account != null) {
+        activeAccount = account;
+        persisteActiveAccount();
```
persistActiveAccount

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -96,42 +111,108 @@ public Shell getShell() {
+  public Set<Account> listAccounts() {
     synchronized (loginState) {
-      loginState.logOut(false /* Don't prompt for logout. */);
+      return accounts;
```
does this expose mutable internal state?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,91 @@
+   *     including failed and canceled login
    */
-  public Credential getActiveCredential(String dialogMessage);
+  public Account logIn(String dialogMessage);
```
we don't use public on interface methods since it's implied

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,91 @@
+   *
+   * @see #logIn
+   */
+  public Account getActiveAccountWithAutoLogin(String dialogMessage);
```
we don't use public on interface methods since it's implied

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,91 @@
    * Safe to call from non-UI contexts.
    */
-  public void clearCredential();
+  public void logOutAll();
```
we don't use public on interface methods since it's implied

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,91 @@
+   * @return true if there existed an account matching the {@code email} and it became active;
+   *     false otherwise
+   */
+  public boolean switchActiveAccount(String email);
```
we don't use public on interface methods since it's implied

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,91 @@
+   *
+   * @return never {@code null}
+   */
+  public Set<Account> listAccounts();
```
we don't use public on interface methods since it's implied

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/messages.properties:
```diff
@@ -1,12 +1,14 @@
-LOGIN_TOOLTIP_LOGGED_OUT=Authorize Eclipse to operate on Google Cloud Platform resources
 LOGOUT_CONFIRM_DIALOG_MESSAGE=Do you want to sign out from Google?
 LOGOUT_CONFIRM_DIALOG_TITILE=Sign out from Google
+MESSAGE_LABEL_ACTIVE_ACCOUNT=Default Account
```
Why MESSAGE_?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,163 @@
+      for (Account account : accounts) {
+        if (!account.getEmail().equals(activeAccount.getEmail())) {
+          Link link = new Link(accountsIndenter, SWT.NO_FOCUS);
+          link.setText("<a href=\"" + account.getEmail() + "\">" + account.getEmail() + "</a>"); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
```
strange use of a element; maybe it makes sense in context of Eclipse

chanseokoh replied:
Yhea, the `Link` widget uses the `<A>` element to mark a link: http://www.java2s.com/Code/Java/SWT-JFace-Eclipse/Createalinkwidget.htm

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,163 @@
+    addAccountButton.addSelectionListener(new LogInOnClick());
+    GridDataFactory.defaultsFor(addAccountButton).applyTo(addAccountButton);
+
+    boolean loggedIn = loginService.getActiveAccount() != null;
```
I've seen this several times now. Maybe worth adding an isLoggedIn method to loginService?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -152,7 +155,7 @@ public Account logIn(String dialogMessage) {
         activeAccount = account;
         accounts = loginState.listAccounts();
-        persisteActiveAccount();
+        persistActiveAccount();
```
did you mean this to be recursive? Could it be an infinite loop?

chanseokoh replied:
This is no recursion!

akerekes commented:
I can see that `Sign Out` will affect all accounts. We may want to make this more fine grained to offer sign out per account.
Also instead of `Other Accounts` label I'd like to suggest `Switch to...` label in the dialog so it may convey better the effect of clicking on of the links.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -60,14 +60,19 @@
+  void logOutAll();
+
+  /**
+   * @return true iff {@link #getActiveAccount} does not return {@code null}.
```
would prefer a less implementation dependent description of what this means

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+        false /* persistSize */,
+        false /* persistLocation */,
+        false /* showDialogMenu */,
+        true /* showPersistActions */,
```
From the Javadoc:
`This parameter has no effect if <code>showDialogMenu</code> is <code>false</code>.`

chanseokoh replied:
Good catch.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+
+  @Override
+  protected Color getBackground() {
+    return getShell().getDisplay().getSystemColor(SWT.COLOR_WIDGET_BACKGROUND);
```
Does this change when the theme changes (e.g. to dark)?

briandealwis replied:
Rather than overriding `getForeground()` and `getBackground()`, call `setForeground()` and `setBackground()` in the constructor.  But I think these overrides are the default colours anyways?

chanseokoh replied:
In fact, the `PopUp` dialog has black background. I'll set these in the constructor.

chanseokoh replied:
(`PopupDialog` doesn't have `setForeground()` and `setBackground()`)

Turns out that `PopupDialog` does `applyColors()` which happens after `createDialogArea()`, so any attempt to override the default colors will need overriding a method anyway. It feels right to override `getBackground()` and `getForeground()`.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+  @Override
+  protected Control createDialogArea(Composite parent) {
+    Composite container = (Composite) super.createDialogArea(parent);
+    GridLayoutFactory.swtDefaults().margins(DEFAULT_MARGIN, DEFAULT_MARGIN).applyTo(container);
```
`GridLayoutFactory.swtDefaults()` uses 5,5 for margin by default.

chanseokoh replied:
:+1:

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+
+  @Override
+  protected Color getForeground() {
+    return getShell().getDisplay().getSystemColor(SWT.COLOR_WIDGET_FOREGROUND);
```
Does this change when the theme changes (e.g. to dark)?

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+
+  private void createButtons(Composite container) {
+    Composite buttonArea = new Composite(container, SWT.NONE);
+    GridDataFactory.defaultsFor(buttonArea).align(SWT.END, SWT.BEGINNING).applyTo(buttonArea);
```
`GridDataFactory.fillDefaults().applyTo(buttonArea)` might achieve the same effect.

chanseokoh replied:
Cool!

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+  private void createButtons(Composite container) {
+    Composite buttonArea = new Composite(container, SWT.NONE);
+    GridDataFactory.defaultsFor(buttonArea).align(SWT.END, SWT.BEGINNING).applyTo(buttonArea);
+    GridLayoutFactory.swtDefaults().numColumns(2).margins(0, 0).applyTo(buttonArea);
```
`GridLayoutFactory.fillDefaults()` uses 0 margins by default.

chanseokoh commented:
Enabling removing (logging out) a single account sounds like a good idea: https://github.com/GoogleCloudPlatform/ide-login/issues/31

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+    Button addAccountButton = new Button(buttonArea, SWT.PUSH);
+    addAccountButton.setText(Messages.BUTTON_ACCOUNTS_PANEL_ADD_ACCOUNT);
+    addAccountButton.addSelectionListener(new LogInOnClick());
+    GridDataFactory.defaultsFor(addAccountButton).applyTo(addAccountButton);
```
Unless you want to change anything on the defaults, GridLayoutFactory.applyTo(parentContainer) usually takes care of the child widgets without the need to set GridData objects explicitly.

briandealwis replied:
I think you're thinking of `generateLayout()`.

akerekes replied:
yes, thanks for the correction!

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+      Button logOutButton = new Button(buttonArea, SWT.PUSH);
+      logOutButton.setText(Messages.BUTTON_ACCOUNTS_PANEL_LOGOUT);
+      logOutButton.addSelectionListener(new LogOutOnClick());
+      GridDataFactory.defaultsFor(logOutButton).applyTo(logOutButton);
```
Unless you want to change anything on the defaults, GridLayoutFactory.applyTo(parentContainer) usually takes care of the child widgets without the need to set GridData objects explicitly.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,162 @@
+      loginService.switchActiveAccount(email);
+
+      close();
+      new AccountsPanel(getParentShell(), loginService).open();
```
Wouldn't it be better to update the dialog instead of recreating it?

chanseokoh replied:
I don't see a particular problem in closing the dialog and opening a new one. I think it's the simplest approach that gets the job done. However, I would seriously considering updating the dialog if you could think of an implementation that is as simple as these two lines (`close()` and `open()`).

akerekes replied:
My concern was around resource consumption and the slightly increased GC-ing. Probably not too much for this rather simple dialog.

chanseokoh commented:
Addressed all the review comments (if I hadn't missed any) and added unit tests for `AccountsPanel`.

I am still working on adding unit tests for `GoogleLoginService`.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/messages.properties:
```diff
@@ -1,12 +1,14 @@
 LOGIN_ERROR_DIALOG_TITLE=Sign in Failed
 LOGIN_ERROR_LOCAL_SERVER_RUN=Error listening for authorization code:\n\n
-LOGIN_MENU_LOGGED_IN=Sign out from Google
+LOGIN_MENU_LOGGED_IN=Sign in to/Sign out from Google...
```
I don't understand this message. It feels like it should be one or the other. If the user can do both, this should be two separate menu items.

chanseokoh replied:
Okay, I tried to implement dynamically switching between one menu item ("sign in...") and two menu items ("add account..." and "sign out...") depending on the login status, but it got a bit ugly. @elharo Do you think having two menu items is a way to go?

briandealwis replied:
+1 to _Manage Google Accounts…_

elharo commented:
Dynamically switching menu items is a UI antipattern because it confuses muscle memory. A non-functional menu item should be disabled, not removed.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,96 @@
-   * @param dialogMessage custom login dialog message. Can be null
+   * @see #logIn
+   */
+  Account getActiveAccountWithAutoLogin(String dialogMessage);
```
I'm not a big fan of this name as _auto-login_ doesn't really correspond to the behaviour.  It would be good to use a different prefix for methods that may block.  Maybe `obtain`?  So `getActiveAccount()` will return null if there's no active account, but `obtainActiveAccount()`?

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,96 @@
+
+import java.util.Set;
 
 public interface IGoogleLoginService {
```
Should have interface javadoc describing the invariants.  E.g., 
  - when there is a logged-in account, one will always be selected as the active account

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,96 @@
+   * @return true if there is at least one signed-in account; false otherwise
    */
-  public Credential getActiveCredential(String dialogMessage);
+  boolean isLoggedIn();
```
This name doesn't seem quite right either: is it that we have an active account, or we have at least one account?  `hasLogins()`?  `hasAccounts()`?  `hasActiveAccount()`?

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,96 @@
+   * @return never {@code null}
    */
-  public void clearCredential();
+  Set<Account> listAccounts();
```
`getAccounts()`?

elharo commented:
As I was discussing in person with @chanseokoh earlier, I'm not sure I like the idea of an active account. It doesn't seem that different from a single account that logs in and out to switch accounts. 

Is it conceivable that we could drop the requirement that the user choose an active account? E.g. deployment works if any of the logged in accounts have the right permissions, not just the active account. 

We can discuss further at our team meeting.

patflynn commented:
I tend to agree although I'm concerned by the statement: " E.g. deployment
works if any of the logged in accounts have the right permissions, not just
the active account."

That sounds wrong to me.  Two users can have access to the same project. A
user needs to be able to clearly indicate as which user they are performing
an action.


On Thu, Sep 29, 2016 at 1:33 PM, Elliotte Rusty Harold <
notifications@github.com> wrote:

> As I was discussing in person with @chanseokoh
> <https://github.com/chanseokoh> earlier, I'm not sure I like the idea of
> an active account. It doesn't seem that different from a single account
> that logs in and out to switch accounts.
>
> Is it conceivable that we could drop the requirement that the user choose
> an active account? E.g. deployment works if any of the logged in accounts
> have the right permissions, not just the active account.
>
> We can discuss further at our team meeting.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#issuecomment-250536423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HeafLEmFLYREewfRYVOWEDJqLJXhks5qu_ZfgaJpZM4KHAwH>
> .
>

chanseokoh commented:
For the record (and a heads-up for @briandealwis), let me summarize the direction after our discussions:
- Remove the activeness notion
- Deploy dialog (at the time of deploy) has a user selector (to be done in this PR)

That's really it. Profit.

elharo commented:
Just capturing the output of our in-person meeting:

1. We are going to remove the notion of an active account (at least insofar as it's visible to the end user.
2. The deploy dialog will contain a user selector drop-down that enables the end user to pick from the logged in accounts.
3. The deploy dialog will remember which account was most recently selected and present that as the selected account in the drop down.

In future PRs we may:

1. Add some sort of "Login to new account" button to the deploy dialog, especially if the user has not logged into any accounts. 
2. Use different user selectors that have different currently selected accounts for operations other than deploy.

chanseokoh commented:
- Added an `AccountSelector`. Deploy works as intended. Users' credential choice is not being validated and saved yet. (#759)

![deployment configuration _017](https://cloud.githubusercontent.com/assets/10523105/19006860/d9fed0d0-872f-11e6-8337-dac1b0a882e1.png)

![properties for nativeproj _016](https://cloud.githubusercontent.com/assets/10523105/19006864/dcdca7be-872f-11e6-94a4-53b0efd99df0.png)

- No active account notion:

![selection_018](https://cloud.githubusercontent.com/assets/10523105/19006943/5ad9febe-8730-11e6-8ba6-aeb07ebefb62.png)

chanseokoh commented:
Will spend some more time on cleaning up the code, but the work is basically done.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties:
```diff
@@ -9,6 +9,8 @@ deploy.login.dialog.message=Browser opened to authorize deployment.
 deploy.manual.link=To manually promote a version use the <a href="{0}">Google Cloud Console.</a>
 deploy.preferences.dialog.title=Deployment configuration
 deploy.preferences.dialog.title.withProject=Deployment configuration for {0}
+deploy.preferences.dialog.label.selectCredential=Credential to use:
```
probably, Credential --> Account
Credential is not a term users think of in this context

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties:
```diff
@@ -9,6 +9,8 @@ deploy.login.dialog.message=Browser opened to authorize deployment.
 deploy.preferences.dialog.title=Deployment configuration
 deploy.preferences.dialog.title.withProject=Deployment configuration for {0}
+deploy.preferences.dialog.label.selectCredential=Credential to use:
+deploy.preferences.dialog.accountSelector.login=<Add a new account...>
```
why the <>?

chanseokoh replied:
I occasionally saw the angle brackets for a special "value" in a combo box in order to distinguish it from other actual values. However, I'll remove them as I don't have any objection doing so.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java:
```diff
@@ -59,7 +59,7 @@
 /**
  * Command handler to deploy an App Engine web application project to App Engine Standard.
```
"an App Engine web application" --> a web application

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -33,64 +40,171 @@
+
+    assertEquals(account1, account);
+    assertTrue(loginService.hasAccounts());
+    // Comparison between Account's is conveniently based only on email. (See 'Account.equals().')
```
accounts

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountSelectorTest.java:
```diff
@@ -0,0 +1,184 @@
+
+    assertEquals(1, selector.combo.getSelectionIndex());
+    assertEquals(emailAtIndex1, selector.combo.getItem(1));
+    switch (emailAtIndex1) {
```
There shouldn't be a switch like this in a test. It means some of your asserts are not being activated. The test should be deterministic and hit all cases.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -64,6 +64,9 @@ public static String getGoogleLoginUrl(String redirectUrl) {
         GoogleLoginService.OAUTH_SCOPES).toString();
   }
 
+  // We expose the reference 'accounts' to callers as-is. 'GoogleLoginService' must not modify
```
This invariant does not seem to be maintained. The logOut method violates it.

chanseokoh replied:
> This invariant does not seem to be maintained. The logOut method violates it.

I'll update the comment to eliminate any possible confusion.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/IGoogleLoginService.java:
```diff
@@ -1,50 +1,61 @@
 
   /**
-   * Clears all credentials. ("logging out" from user perspective.)
+   * Returns a list of currently logged-in accounts.
```
delete "a list of"

patflynn commented:
+1 to not using the credential term. A simple 'Account:' label would do IMO

On Fri, Sep 30, 2016 at 6:41 PM, Elliotte Rusty Harold <
notifications@github.com> wrote:

> *@elharo* commented on this pull request.
> ------------------------------
>
> In plugins/com.google.cloud.tools.eclipse.appengine.
> deploy.ui/src/com/google/cloud/tools/eclipse/appengine/
> deploy/ui/messages.properties
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>
> :
>
> > @@ -9,6 +9,8 @@ deploy.login.dialog.message=Browser opened to authorize deployment.
>  deploy.manual.link=To manually promote a version use the <a href="{0}">Google Cloud Console.</a>
>  deploy.preferences.dialog.title=Deployment configuration
>  deploy.preferences.dialog.title.withProject=Deployment configuration for {0}
> +deploy.preferences.dialog.label.selectCredential=Credential to use:
>
> probably, Credential --> Account
> Credential is not a term users think of in this context
> ------------------------------
>
> In plugins/com.google.cloud.tools.eclipse.appengine.
> deploy.ui/src/com/google/cloud/tools/eclipse/appengine/
> deploy/ui/messages.properties
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>
> :
>
> > @@ -9,6 +9,8 @@ deploy.login.dialog.message=Browser opened to authorize deployment.
>  deploy.manual.link=To manually promote a version use the <a href="{0}">Google Cloud Console.</a>
>  deploy.preferences.dialog.title=Deployment configuration
>  deploy.preferences.dialog.title.withProject=Deployment configuration for {0}
> +deploy.preferences.dialog.label.selectCredential=Credential to use:
> +deploy.preferences.dialog.accountSelector.login=<Add a new account...>
>
> why the <>?
> ------------------------------
>
> In plugins/com.google.cloud.tools.eclipse.appengine.
> deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/
> StandardDeployCommandHandler.java
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>
> :
>
> > @@ -59,7 +59,7 @@
>  /**
>   * Command handler to deploy an App Engine web application project to App Engine Standard.
>
> "an App Engine web application" --> a web application
> ------------------------------
>
> In plugins/com.google.cloud.tools.eclipse.appengine.login.
> test/src/com/google/cloud/tools/eclipse/appengine/login/
> GoogleLoginServiceTest.java
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>
> :
>
> >      GoogleLoginService loginService = new GoogleLoginService(dataStore, uiFacade, loggerFacade);
> -    Assert.assertNull(loginService.getCachedActiveCredential());
> +    assertTrue(loginService.getAccounts().isEmpty());
> +  }
> +
> +  @Test
> +  public void testLogIn_successfulLogin() {
> +    GoogleLoginService loginService = newLoginServiceWithMockLoginState(true /* set up logins */);
> +    Account account = loginService.logIn(null /* no dialog message */);
> +
> +    assertEquals(account1, account);
> +    assertTrue(loginService.hasAccounts());
> +    // Comparison between Account's is conveniently based only on email. (See 'Account.equals().')
>
> accounts
> ------------------------------
>
> In plugins/com.google.cloud.tools.eclipse.appengine.login.
> test/src/com/google/cloud/tools/eclipse/appengine/login/
> ui/AccountSelectorTest.java
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>
> :
>
> > +  @Test
> +  public void testGetSelectedCredential() {
> +    when(loginService.getAccounts())
> +        .thenReturn(new HashSet<>(Arrays.asList(account1, account2, account3)));
> +    AccountSelector selector = new AccountSelector(shell, loginService, "<select this to login>");
> +
> +    assertEquals(-1, selector.combo.getSelectionIndex());
> +    assertNull(selector.getSelectedCredential());
> +    String emailAtIndex1 = selector.combo.getItem(1);
> +
> +    selector.combo.select(1);
> +    selector.logInOnSelect.widgetSelected(mock(SelectionEvent.class));
> +
> +    assertEquals(1, selector.combo.getSelectionIndex());
> +    assertEquals(emailAtIndex1, selector.combo.getItem(1));
> +    switch (emailAtIndex1) {
>
> There shouldn't be a switch like this in a test. It means some of your
> asserts are not being activated. The test should be deterministic and hit
> all cases.
> ------------------------------
>
> In plugins/com.google.cloud.tools.eclipse.appengine.login/
> src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>
> :
>
> > @@ -64,6 +64,9 @@ public static String getGoogleLoginUrl(String redirectUrl) {
>          GoogleLoginService.OAUTH_SCOPES).toString();
>    }
>
> +  // We expose the reference 'accounts' to callers as-is. 'GoogleLoginService' must not modify
>
> This invariant does not seem to be maintained. The logOut method violates
> it.
> ------------------------------
>
> In plugins/com.google.cloud.tools.eclipse.appengine.login/
> src/com/google/cloud/tools/eclipse/appengine/login/
> IGoogleLoginService.java
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>
> :
>
> >
>    /**
> -   * Clears all credentials. ("logging out" from user perspective.)
> +   * Returns a list of currently logged-in accounts.
>
> delete "a list of"
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742#pullrequestreview-2412679>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HfmJLf33tekMAT4LEUdcLRi5LA5Kks5qvZADgaJpZM4KHAwH>
> .
>

chanseokoh commented:
Ready for review. PTAL.

chanseokoh commented:
Builds are failing with Tycho where they shouldn't: `Caused by: java.util.zip.ZipException: archive is not a ZIP archive`

Should be a transient error, but it's consistently happening at the moment.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties:
```diff
@@ -9,6 +9,8 @@ deploy.login.dialog.message=Browser opened to authorize deployment.
 deploy.manual.link=To manually promote a version use the <a href="{0}">Google Cloud Console.</a>
 deploy.preferences.dialog.title=Deployment configuration
 deploy.preferences.dialog.title.withProject=Deployment configuration for {0}
+deploy.preferences.dialog.label.selectAccount=Account to use:
```
Maybe just "Account" because "to use" is implied

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java:
```diff
@@ -57,9 +58,9 @@
 
 /**
- * Command handler to deploy an App Engine web application project to App Engine Standard.
+ * Command handler to deploy an web application project to App Engine Standard.
```
a web

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java:
```diff
@@ -82,10 +83,19 @@ public Object execute(ExecutionEvent event) throws ExecutionException {
+          } else {
+            // Stopgap (https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/759)
+            MessageDialog.openInformation(HandlerUtil.getActiveShell(event),
+                "Account Required for Deploy", "Please select an account to deploy.");
```
move these strings to a message bundle

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/messages.properties:
```diff
@@ -1,12 +1,12 @@
-LOGIN_MENU_LOGGED_IN=Sign out from Google
+LOGIN_MENU_LOGGED_IN=Manage Google Accounts...
 LOGIN_MENU_LOGGED_OUT=Sign in to Google...
 LOGIN_PROGRESS_DIALOG_MESSAGE=Please complete signing in using the external browser.
```
delete "Please" http://ux.stackexchange.com/questions/25500/why-avoid-the-word-please-in-error-messages

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties:
```diff
@@ -24,4 +24,8 @@ use.custom.versioning=Override generated version
 action.remove=Remove
-job.terminated.template=<terminated> {0}
\ No newline at end of file
+job.terminated.template=<terminated> {0}
```
why <>?

chanseokoh replied:
I didn't touch the message. @akerekes may have put the angle brackets, and I think they are suitable for console messages.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties:
```diff
@@ -24,4 +24,8 @@ use.custom.versioning=Override generated version
+
+# TODO(chanseok): remove after fixing https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/759.)
+deploy.account.missing.dialog.title=Account Required for Deploy
+deploy.account.missing.dialog.message=Please try again after selecting an account.
```
No "Please"

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -33,64 +40,171 @@
 
   @Test
   public void testGoogleLoginService_restoreSavedCredential() {
     // Persisted credential in the data store is valid.
```
Nit: "persisted" is redundant with "in the data store". Suggest just "Credential in the data store is valid"

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -33,64 +40,171 @@
+
+  @Test
+  public void testGoogleLoginService_removeSavedCredentialIfScopesChanged() {
     // Persisted credential in the data store has an out-dated scopes.
```
Nit: "Persisted" is redundant with "in the data store". Suggest just "Credential in the data store"

elharo replied:
ping

chanseokoh replied:
This had been fixed.

chanseokoh commented:
I have retried many many times, but build always fails on `orackejdk8`. Needs some investigation.

```
$ mvn -Ptravis --fail-at-end verify ${ECLIPSE_TARGET:+-Declipse.target=${ECLIPSE_TARGET}}
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=192m; support was removed in 8.0
[INFO] Scanning for projects...
[WARNING] Error initializing: org.eclipse.tycho.p2.resolver.P2DependencyResolver@4287d447
java.lang.RuntimeException: org.apache.maven.MavenExecutionException: Failed to unpack Tycho's OSGi runtime: Error while expanding /home/travis/.m2/repository/org/eclipse/tycho/tycho-bundles-external/0.24.0/tycho-bundles-external-0.24.0.zip
```

```
Caused by: org.apache.maven.MavenExecutionException: Failed to unpack Tycho's OSGi runtime: Error while expanding /home/travis/.m2/repository/org/eclipse/tycho/tycho-bundles-external/0.24.0/tycho-bundles-external-0.24.0.zip
```

```
Caused by: org.codehaus.plexus.archiver.ArchiverException: Error while expanding /home/travis/.m2/repository/org/eclipse/tycho/tycho-bundles-external/0.24.0/tycho-bundles-external-0.24.0.zip
```

```
Caused by: java.util.zip.ZipException: archive is not a ZIP archive
```

chanseokoh commented:
I cleared the Travis cache for this PR, and the problem seems to be gone.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,137 @@
+ *
+ * TODO(chanseok): move the panel to the bottom-right corner: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/740
+ */
+public class AccountsPanel extends PopupDialog {
```
I think you need to override getInitialLocation() and return a `Point` that's near the mouse cursor.  Use `Display#getCursorLocation()`.  See `Window#getInitialLocation()` for how it uses the `Monitor` to constrain the location so the contents aren't shown off-screen.

chanseokoh replied:
Closed #740, so no longer relevant.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,137 @@
+        false /* showPersistActions */,
+        null /* no title area */, null /* no info text area */);
+    this.loginService = loginService;
+  }
```
I think this dialog would benefit from:

```
    setTitleText("Google Accounts");
    setInfoText(MessageFormat.format("Click outside or hit {0} to dismiss", SWTKeyLookup.ESC_NAME));
```
(or something along that lines)

akerekes replied:
I wouldn't add that info text. In Eclipse many popups work this way, so I'd think users expect it.

briandealwis replied:
Ok, we can always add it if we get reports of confusion.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties:
```diff
@@ -22,4 +24,8 @@ use.custom.versioning=Override generated version
+job.terminated.template=<terminated> {0}
+
+# TODO(chanseok): remove after fixing https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/759.)
+deploy.account.missing.dialog.title=Account Required for Deploy
```
https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/759

elharo replied:
so this message shouldn't be needed, right?

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/messages.properties:
```diff
@@ -1,12 +1,12 @@
+BUTTON_ACCOUNTS_PANEL_ADD_ACCOUNT=Add Account
+BUTTON_ACCOUNTS_PANEL_LOGOUT=Sign Out
```
You might want to try using `SWT.FLAT` in your buttons.  They're smaller and look a bit better in the dialog IMHO.

chanseokoh replied:
Good idea. Used `SWT.FLAT`.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java:
```diff
@@ -0,0 +1,137 @@
+ * A panel listing all currently logged-in accounts. The panel allows adding new accounts and
+ * logging out all accounts.
+ *
+ * TODO(chanseok): move the panel to the bottom-right corner: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/740
```
Is this still relevant? Looks like that issue is closed.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties:
```diff
@@ -9,6 +9,8 @@ deploy.login.dialog.message=Browser opened to authorize deployment.
 deploy.preferences.dialog.title=Deployment configuration
 deploy.preferences.dialog.title.withProject=Deployment configuration for {0}
+deploy.preferences.dialog.label.selectAccount=Account:
+deploy.preferences.dialog.accountSelector.login=Add a new account...
```
This is an item in the combo box, and I recall I have seen many occasions where ankle brackets are used to distinguish this kind of instructive item from actual values in the combo list. I think this is not unusual as @briandealwis also suggested it.

akerekes commented:
I used the `<>` because I wanted to follow the format Eclipse uses in the
console when a process has terminated.

On Tue, Oct 4, 2016, 16:04 Chanseok Oh <notifications@github.com> wrote:

> *@chanseokoh* commented on this pull request.
> ------------------------------
>
> In
> plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742>:
>
> > \ No newline at end of file
> +job.terminated.template=<terminated> {0}
>
> I didn't touch the message. @akerekes <https://github.com/akerekes> may
> have put the angle brackets, and I think they are suitable for console
> messages.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/742>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADOO_uj6BdpeTKvefGCSN_4J2gKVrFulks5qwrFLgaJpZM4KHAwH>
> .
>

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -97,41 +102,56 @@ public Shell getShell() {
   @VisibleForTesting
-  GoogleLoginService(
+  GoogleLoginService(OAuthDataStore dataStore, LoginServiceUi uiFacade, LoggerFacade loggerFacade) {
+    this(new GoogleLoginState(Constants.getOAuthClientId(), Constants.getOAuthClientSecret(),
```
I have no idea what is meant by an emoji here. is this addressed?

chanseokoh replied:
The formatting in my opinion is correct.

elharo commented:
please merge
